Using installed version OpenCASCADE

katyo opened the pull request:
This PR adds support installed version of **OpenCASCADE** library when it found and fits prerequisites. This drastically speeds up build. You can use `builtin` feature to prefer builtin version and skip detecting installed one.

Also added support for dynamic linking via using `dynamic` feature so this PR also superseeds #26.

katyo commented:
@bschwind What do you think?

bschwind commented:
Hey @katyo, thanks for the PR! I think once we figure out where opencascade-sys is going (being discussed in #86), we can then add these changes on top (or vice-versa).

I should also see if this helps resolve #27.

In any case, please give me a bit of time to test and review :)

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -1,10 +1,94 @@
+    }
+
+    // Try find packaged version of OpenCASCADE or use builtin
+    let occ_config = if is_builtin { None } else { find_opencascade(is_dynamic, OCC_VERSION) }
+        .unwrap_or_else(|| build_opencascade(is_dynamic));
```
This falls back to the bundled OpenCascade if `dynamic` is set but we cannot find it. That feels a bit too magic/surprising - I think it would be better to fail in this case, tell the user to try changing cargo flags instead.

katyo replied:
I tried preserve current behavior as much as possible so I think the falling back to builtin version is a right thing.

strohel replied:
> I tried preserve current behavior as much as possible

Oh yeah that's a good point, we don't want to break the build for people, and using the built-in OCC is more universal.

However I think that should be solved by making the `builtin` feature flag enabled by default on `opencascade` level.

The automatic fallback could be very confusing: imagine people disabling the `builtin` flag and still getting horrible build times (or opencascade-rs not picking up their custom-patched system-installed OCC) - perhaps they've installed OCC system-wide, but they version is not compatible. They would not be able to know the fact because the build won't tell them (output of build scripts is only visible with `cargo ... -vv`).

strohel reviewed (changes requested):
Thanks @katyo, this is indeed a very useful change. I posted some thoughts/suggestions for the implementation. We've also working on https://github.com/bschwind/opencascade-rs/pull/86 with @bschwind. I'm happy to rebase/adapt this if #86 lands sooner.

BTW this could be even a solution for slow CI - at least the Ubuntu runner provides opencascade in the repository.

katyo commented on crates/opencascade-sys/OCCT_PKG/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required (VERSION 3.1 FATAL_ERROR)
+project (OpenCASCADEPackageConfig)
+find_package (OpenCASCADE REQUIRED)
```
Course, I already used such trick before.

strohel commented on crates/opencascade-sys/Cargo.toml:
```diff
@@ -46,6 +46,10 @@ exclude = [
 
+[features]
+dynamic = []
+builtin = []
```
Ah yeah, this is a good approach (If I understand it correctly):
- `builtin` -> build&link OCC statically
- `!builtin` -> use whatever type of linking that system OCC uses

katyo commented:
@bschwind I rebased this PR to current master.
@strohel I changed it as you suggested before but I have some doubts.

Maybe we should enable `builtin` feature by default or remove it in favor of `external` feature which will enable using installed OpenCASCADE libraries instead of compiling builtin one. But it seems cargo does not provide a way to disable dependency conditionally when some feature is enabled.

bschwind commented:
I've been delayed in reviewing this but I'll give it a try soon!

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -1,35 +1,45 @@
-use occt_sys::{occt_include_path, occt_lib_path};
+const OCCT_VERSION: (u8, u8) = (7, 2);
```
Nit: it would be nice to document that this is a semver requirement.

bschwind replied:
This might need to be moved into the relevant `detect()` function to avoid an unused const warning from clippy.

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -54,3 +64,65 @@ fn main() {
+        let dst = std::panic::catch_unwind(|| cmake::Config::new("OCCT_PKG").build())
+            .expect("Pre-installed OpenCASCADE library not found! You can use `builtin` feature if you do not want install OCCT libraries system-wide.");
+        let cfg = std::fs::read_to_string(dst.join("share").join("occ_info.txt"))
+            .expect("Something went wrong when detecting pre-installed OpenCASCADE library!");
```
```suggestion
            .expect("Something went wrong when detecting pre-installed OpenCASCADE library.");
```

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -54,3 +64,65 @@ fn main() {
+            (version_major, version_minor, include_dir, library_dir)
+        {
+            if version_major != OCCT_VERSION.0 {
+                panic!("Pre-installed OpenCASCADE library found but major version is not the same (found {}.x but {}.x required)! Please provide required version or use `builtin` feature.", version_major, OCCT_VERSION.0);
```
```suggestion
                panic!("Pre-installed OpenCASCADE library found but major version is not the same (found {}.x but {}.x required). Please provide required version or use `builtin` feature.", version_major, OCCT_VERSION.0);
```

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -54,3 +64,65 @@ fn main() {
+
+            if version_minor < OCCT_VERSION.1 {
+                panic!(
+                    "Pre-installed OpenCASCADE library found but minor version is lower than required ({}.{} < {}.{})! Please provide required version or use `builtin` feature.",
```
```suggestion
                    "Pre-installed OpenCASCADE library found but minor version is lower than required ({}.{} < {}.{}). Please provide required version or use `builtin` feature.",
```

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -54,3 +64,65 @@ fn main() {
+        Self { include_dir, library_dir, is_dynamic: false }
+    }
+
+    #[cfg(not(feature = "builtin"))]
+    fn default() -> Self {
+        let dst = std::panic::catch_unwind(|| cmake::Config::new("OCCT_PKG").build())
+            .expect("Pre-installed OpenCASCADE library not found! You can use `builtin` feature if you do not want install OCCT libraries system-wide.");
```
```suggestion
            .expect("Pre-installed OpenCASCADE library not found. You can use `builtin` feature if you do not want to install OCCT libraries system-wide.");
```

strohel replied:
Supernit: I think it is customary that `Default::default()` is a cheap method. This one is heavyweight (builds the stub cmake project), so we could perhaps make these inherent methods called `fn detect() -> Self` or similar.

strohel commented on crates/opencascade-sys/build.rs:
```diff
@@ -54,3 +64,65 @@ fn main() {
+}
+
+impl Default for OcctConfig {
+    #[cfg(feature = "builtin")]
+    fn default() -> Self {
+        let include_dir = occt_sys::occt_include_path();
+        let library_dir = occt_sys::occt_lib_path();
+
+        Self { include_dir, library_dir, is_dynamic: false }
+    }
```
Optional: it could be nice to check `OCCT_VERSION` also against the bundled version so that we don't forget to update other if we update of these.

katyo replied:
I thought about it too but for now I haven't any ideas how to implement it.

strohel reviewed (commented):
Looking good! Just adding some nits.

> Maybe we should enable builtin feature by default

Yes, please, on `opencascade` level.

bschwind reviewed (commented):
This is looking good, just need a few tweaks:

* Fix the build errors
* Fix the clippy warnings (we run `cargo clippy --release --all-targets -- -D warnings` in CI)

Could you also add a bit of documentation in the README on how to use the bundled OCCT, vs. using the system-installed one?

katyo commented:
@bschwind @strohel I thought a more proper solution is using cmake in both cases (when `builtin` feature is enabled and also when it is disabled).

bschwind commented:
@katyo could you please rebase this on the latest `main` (which is now based on opencascade 7.7.1)?

katyo commented:
@bschwind Does this mean that OCCT 7.7.1 is a minimum supported version?

katyo commented:
@bschwind Currently CI won't work as expected when occt-sys needs to be updated too. Hope such behavior was predictable.

bschwind commented:
@katyo that's okay! I see what you mean, we need to update `occt-sys` first with the changes, and _then_ publish that to 0.3.

> @bschwind Does this mean that OCCT 7.7.1 is a minimum supported version?

Unfortunately yes. We could go back and see when exactly the `ComputeNormals` function was added and use that version I suppose, but it was a nightmare to get the crate size down to 10MB, and that later version was needed so I didn't have to depend on the `StdPrs` package.

I can help to support this PR and get it through, as I currently have the publishing rights to `occt-sys`, just give me some time to complete that.

katyo commented:
@bschwind Any chance that this PR will be merged soon?
Seems currently development stuck. I would like add many more changes.

bschwind commented:
Hi @katyo , yes I still want to get to this for sure. I've been on a two week trip so development and review has been a bit slow. This PR is the next one I want to test so I'll update you here when I've done so. Sorry for the wait!

bschwind commented on crates/opencascade-sys/build.rs:
```diff
@@ -52,3 +66,69 @@ fn main() {
+        // Add path to builtin OCCT
+        #[cfg(feature = "builtin")]
+        {
+            std::env::set_var("DEP_OCCT_ROOT", occt_sys::occt_path().as_os_str());
```
Is this environment variable doing anything? I see [documentation](https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html) with env vars of the form `<PACKAGE_NAME>_ROOT`, but not `DEP_<PACKAGE_NAME>_ROOT`

katyo replied:
This is cmake-rs specific thing.
See here [cmake-rs/src/lib.rs#L553](https://github.com/rust-lang/cmake-rs/blob/3ddc3d505ef38ebec54b999110f9dd10b4465252/src/lib.rs#L553)

bschwind commented:
@strohel could I get another review from you on this? Especially concerning the newly added `builtin` feature flags on the `examples` and `viewer` crates, I wasn't sure if that was the right approach or not.

bschwind commented:
@katyo I'm back from the trip I was on. I pushed some changes to your branch and published 0.3 of occt-sys to get this building. Thanks so much for your patience! Please take a look at the changes and let me know if you agree with them.

I've tested this fairly thoroughly and things are working well!

Interestingly, I seem to get better performance when using the builtin version of OCCT. Anecdotally on my mac mini m1, the keycap model takes:

* `~1.58 seconds` using the statically-linked OCCT
* `~2.01 seconds` using the dynamically-linked, system-provided OCCT

katyo commented:
@bschwind 

Are you sure that this works as expected?
Because I tested it with linux only maybe something went wrong.
Please look at `ldd` output on both cases to check which OCCT libs actually used.

bschwind commented:
@katyo I tested it with `--no-default-features` and without `--no-default-features`:

### Dynamic Linking
`$ cargo build --release --no-default-features --bin viewer`

```
opencascade $ otool -L target/release/viewer                          
target/release/viewer:
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 258.17.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2113.20.111)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 56.0.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1557.3.2)
	/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo (compatibility version 1.2.0, current version 1.5.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1856.105.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1856.105.0)
	/opt/homebrew/opt/opencascade/lib/libTKMath.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKernel.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKFeat.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKGeomBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKG2d.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKG3d.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKTopAlgo.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKGeomAlgo.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKBRep.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKPrim.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKSTEP.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKSTEPAttr.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKSTEPBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKSTEP209.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKSTL.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKMesh.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKShHealing.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKFillet.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKBool.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKBO.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKOffset.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/opt/homebrew/opt/opencascade/lib/libTKXSBase.7.7.dylib (compatibility version 7.7.0, current version 7.7.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1200.3.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/System/Library/Frameworks/ColorSync.framework/Versions/A/ColorSync (compatibility version 1.0.0, current version 4.7.0)
```

### Static Linking
`$ cargo build --release  --bin viewer`

```
opencascade $ otool -L target/release/viewer 
target/release/viewer:
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
	/System/Library/Frameworks/Metal.framework/Versions/A/Metal (compatibility version 1.0.0, current version 258.17.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2113.20.111)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 56.0.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1557.3.2)
	/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo (compatibility version 1.2.0, current version 1.5.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1856.105.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1856.105.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1200.3.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/System/Library/Frameworks/ColorSync.framework/Versions/A/ColorSync (compatibility version 1.0.0, current version 4.7.0)
```

katyo commented:
@bschwind Seems it works as expected.

bschwind commented:
@katyo were you having trouble making it work on linux? I can test it out on a linux machine today too.

katyo commented:
@bschwind No, it works fine for me (NixOS). I think it shouldn't have problems with other distros either.

Maybe we won't get significant speedup using dynamic linking itself but using pre-installed OCCT allows skip building it which takes several minutes on my laptop.

bschwind commented:
@strohel I'm going to merge this, but please leave any post-merge reviews you may have and we can always open up new PRs to make any adjustments you might suggest.

@katyo thanks again for your work and patience on this.
